contrib/babelfishpg_tsql/src/pl_handler.c

```diff
@@ -3652,76 +3674,105 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
+										int16 old_priv = get_bbf_schema_privilege(logical_schema, obj,rol_spec->rolename);
+										if(old_priv!= 31 || ((old_priv)&(~(31))) != old_priv)
+											update_bbf_schema_entry(logical_schema, obj, 31, old_priv, rol_spec->rolename, obj_type, grant->is_grant);
 									}
 									if (has_schema_perms)
 										return;
 								}
 								break;
 							}
 						}
-						foreach(lc1, grant->privileges)
+						
+						if (grant->is_grant)
 						{
-							AccessPriv *ap = (AccessPriv *) lfirst(lc1);
-							if (grant->is_grant)
+							/*
+							* 1. Execute the GRANT statement.
+							* 2. Add its corresponding entry in the catalog, if doesn't exist already.
+							* 3. Don't add an entry, if the permission is granted on column list.
+							*/
+							if (prev_ProcessUtility)
+								prev_ProcessUtility(pstmt, queryString, readOnlyTree, context, params,
+									queryEnv, dest, qc);
+							else
+								standard_ProcessUtility(pstmt, queryString, readOnlyTree, context, params,
+									queryEnv, dest, qc);
+							foreach(lc, grant->grantees)
 							{
-								/*
-								* 1. Execute the GRANT statement.
-								* 2. Add its corresponding entry in the catalog, if doesn't exist already.
-								* 3. Don't add an entry, if the permission is granted on column list.
-								*/
-								if (prev_ProcessUtility)
-									prev_ProcessUtility(pstmt, queryString, readOnlyTree, context, params,
-										queryEnv, dest, qc);
-								else
-									standard_ProcessUtility(pstmt, queryString, readOnlyTree, context, params,
-										queryEnv, dest, qc);
-								foreach(lc, grant->grantees)
+								RoleSpec	   *rol_spec = (RoleSpec *) lfirst(lc);
+								bool flag = false;
+								foreach(lc1, grant->privileges)
 								{
-									RoleSpec	   *rol_spec = (RoleSpec *) lfirst(lc);
-									if ((ap->cols == NULL) && (rol_spec->rolename != NULL) && !check_bbf_schema_for_entry(logical_schema, obj, ap->priv_name, rol_spec->rolename))
-										add_entry_to_bbf_schema(logical_schema, obj, ap->priv_name, rol_spec->rolename, obj_type);
+									AccessPriv *ap = (AccessPriv *) lfirst(lc1);
+									if(ap->cols != NULL)
+									{
+										flag = true;
+										break;
+									}
+								}
+								if(flag)
+									break;
+								if ((rol_spec->rolename != NULL) && (strcmp(rol_spec->rolename, "public") != 0) && !check_bbf_schema_for_entry(logical_schema, obj, rol_spec->rolename))
+									add_entry_to_bbf_schema(logical_schema, obj, privilege_maskInt, rol_spec->rolename, obj_type);
+								else if((rol_spec->rolename != NULL) && (strcmp(rol_spec->rolename, "public") != 0))
+								{
+									int16 old_priv = get_bbf_schema_privilege(logical_schema, obj, rol_spec->rolename);
+									if(old_priv!= privilege_maskInt || ((old_priv)|(privilege_maskInt)) != old_priv)
+										update_bbf_schema_entry(logical_schema, obj, privilege_maskInt, old_priv, rol_spec->rolename, obj_type, grant->is_grant);
```

Comment:
This code appears in 7-8 places in the code. Please generalize and decouple this in a separate function (probably inline).